File: src/ideago/models/research.py
"""Research domain models.

竞品调研领域模型：Platform、RawResult、Intent、Competitor、SourceResult、ResearchReport。
"""

import hashlib
from datetime import datetime, timezone
from enum import Enum
from typing import Any
from uuid import uuid4

from pydantic import Field

from ideago.models.base import BaseModel, TimestampMixin


class Platform(str, Enum):
    """Supported data source platforms / 支持的数据源平台。"""

    GITHUB = "github"
    TAVILY = "tavily"
    HACKERNEWS = "hackernews"
    APPSTORE = "appstore"
    PRODUCT_HUNT = "producthunt"
    GOOGLE_TRENDS = "google_trends"


class RawResult(BaseModel):
    """Single raw result from a data source / 数据源返回的单条原始结果。"""

    title: str = Field(description="Result title / 结果标题")
    description: str = Field(default="", description="Result description / 结果描述")
    url: str = Field(description="Source URL, mandatory / 来源链接（必填）")
    platform: Platform = Field(description="Source platform / 来源平台")
    raw_data: dict[str, Any] = Field(
        default_factory=dict,
        description="Raw API response preserved for debugging / 原始响应备份",
    )
    fetched_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        description="Fetch timestamp / 抓取时间",
    )


class SearchQuery(BaseModel):
    """Platform-specific search queries / 平台定制搜索词。"""

    platform: Platform = Field(description="Target platform / 目标平台")
    queries: list[str] = Field(
        min_length=1,
        description="Search query strings tailored for this platform",
    )


class Intent(BaseModel):
    """Parsed user intent with per-platform search queries / 解析后的用户意图。"""

    keywords_en: list[str] = Field(
        min_length=1,
        description="English keywords extracted from user input",
    )
    keywords_zh: list[str] = Field(
        default_factory=list,
        description="Chinese keywords if applicable",
    )
    app_type: str = Field(
        description="App form: web / mobile / browser-extension / cli / api / desktop",
    )
    target_scenario: str = Field(
        description="One-sentence target scenario description",
    )
    search_queries: list[SearchQuery] = Field(
        description="Per-platform tailored search queries",
    )
    cache_key: str = Field(
        default="",
        description="Normalized cache key derived from sorted keywords + app_type",
    )

    def compute_cache_key(self) -> str:
        """Generate a deterministic cache key from keywords + app_type.

        Same keywords in different order produce the same key.
        """
        normalized = sorted(k.lower().strip() for k in self.keywords_en)
        raw = f"{self.app_type.lower()}::{'|'.join(normalized)}"
        return hashlib.sha256(raw.encode()).hexdigest()[:16]


class Competitor(BaseModel):
    """A competitor product identified during research / 调研中发现的竞品。"""

    name: str = Field(description="Product or project name / 产品名称")
    links: list[str] = Field(
        min_length=1,
        description="URLs — at least 1 required, no link = not recorded",
    )
    one_liner: str = Field(description="One-sentence positioning / 一句话定位")
    features: list[str] = Field(
        default_factory=list,
        description="Key features / 主要功能",
    )
    pricing: str | None = Field(
        default=None,
        description="Pricing info if available / 定价信息",
    )
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    relevance_score: float = Field(
        default=0.5,
        ge=0.0,
        le=1.0,
        description="0-1 relevance score, higher = more relevant",
    )
    source_platforms: list[Platform] = Field(
        description="Platforms where this competitor was found",
    )
    source_urls: list[str] = Field(
        description="Original pages where info was extracted from",
    )


class RecommendationType(str, Enum):
    """Structured recommendation type / 结构化推荐类型。"""

    GO = "go"
    CAUTION = "caution"
    NO_GO = "no_go"


class SourceStatus(str, Enum):
    """Status of a data source query / 数据源查询状态。"""

    OK = "ok"
    FAILED = "failed"
    CACHED = "cached"
    TIMEOUT = "timeout"
    DEGRADED = "degraded"


class SourceResult(BaseModel):
    """Result from one data source including status / 单个数据源结果（含状态）。"""

    platform: Platform
    status: SourceStatus
    raw_count: int = Field(
        default=0,
        description="Number of raw results fetched / 抓取到的原始结果数",
    )
    competitors: list[Competitor] = Field(default_factory=list)
    error_msg: str | None = Field(default=None)
    duration_ms: int = Field(
        default=0,
        description="Time taken for this source in milliseconds",
    )


class ConfidenceMetrics(BaseModel):
    """Confidence metrics for report quality and source reliability."""

    sample_size: int = Field(default=0, ge=0)
    source_coverage: int = Field(default=0, ge=0)
    source_success_rate: float = Field(default=0.0, ge=0.0, le=1.0)
    freshness_hint: str = Field(default="Generated moments ago")
    score: int = Field(default=0, ge=0, le=100)


class EvidenceItem(BaseModel):
    """Single evidence item attached to report conclusion."""

    title: str = Field(default="")
    url: str = Field(default="")
    platform: str = Field(default="")
    snippet: str = Field(default="")


class EvidenceSummary(BaseModel):
    """Evidence summary section for transparency."""

    top_evidence: list[str] = Field(default_factory=list)
    evidence_items: list[EvidenceItem] = Field(default_factory=list)


class CostBreakdown(BaseModel):
    """Cost and latency telemetry of one report run."""

    llm_calls: int = Field(default=0, ge=0)
    llm_retries: int = Field(default=0, ge=0)
    endpoint_failovers: int = Field(default=0, ge=0)
    source_calls: int = Field(default=0, ge=0)
    pipeline_latency_ms: int = Field(default=0, ge=0)
    tokens_prompt: int = Field(default=0, ge=0)
    tokens_completion: int = Field(default=0, ge=0)


class LlmFaultToleranceMeta(BaseModel):
    """LLM fault-tolerance metadata for this report."""

    fallback_used: bool = Field(default=False)
    endpoints_tried: list[str] = Field(default_factory=list)
    last_error_class: str = Field(default="")


class ReportMeta(BaseModel):
    """Additional report metadata for observability and debugging."""

    llm_fault_tolerance: LlmFaultToleranceMeta = Field(
        default_factory=LlmFaultToleranceMeta
    )
    quality_warnings: list[str] = Field(
        default_factory=list,
        description="Quality guardrail notes for recommendation reliability.",
    )


class ResearchReport(TimestampMixin):
    """Complete research report / 完整调研报告。"""

    id: str = Field(
        default_factory=lambda: str(uuid4()),
        description="Unique report ID / 报告唯一标识",
    )
    query: str = Field(
        description="User's original natural language input / 用户原始输入"
    )
    intent: Intent
    source_results: list[SourceResult] = Field(default_factory=list)
    competitors: list[Competitor] = Field(
        default_factory=list,
        description="Globally deduplicated competitor list / 全局去重竞品列表",
    )
    market_summary: str = Field(
        default="",
        description="LLM-generated market analysis / 市场分析摘要",
    )
    go_no_go: str = Field(
        default="",
        description="Go/No-Go recommendation with reasoning / 推荐建议",
    )
    recommendation_type: RecommendationType = Field(
        default=RecommendationType.GO,
        description="Structured recommendation: go / caution / no_go",
    )
    differentiation_angles: list[str] = Field(
        default_factory=list,
        description="Suggested differentiation points / 差异化切入点",
    )
    confidence: ConfidenceMetrics = Field(default_factory=ConfidenceMetrics)
    evidence_summary: EvidenceSummary = Field(default_factory=EvidenceSummary)
    cost_breakdown: CostBreakdown = Field(default_factory=CostBreakdown)
    report_meta: ReportMeta = Field(default_factory=ReportMeta)

File: frontend/src/lib/types/research.ts
export type Platform =
  | 'github'
  | 'tavily'
  | 'hackernews'
  | 'appstore'
  | 'producthunt'
  | 'reddit'
  | 'google_trends'

export type SourceStatus = 'ok' | 'failed' | 'cached' | 'timeout' | 'degraded'

export type EventType =
  | 'intent_started'
  | 'intent_parsed'
  | 'query_planning_started'
  | 'query_planning_completed'
  | 'source_started'
  | 'source_completed'
  | 'source_failed'
  | 'extraction_started'
  | 'extraction_completed'
  | 'aggregation_started'
  | 'aggregation_completed'
  | 'report_ready'
  | 'cancelled'
  | 'error'

export interface PipelineEventData {
  app_type?: string
  keywords?: string[]
  target_scenario?: string
  platform?: string
  count?: number
  families?: string[]
}

export interface PipelineEvent {
  type: EventType
  stage: string
  message: string
  data: PipelineEventData
  timestamp: string
}

export interface ProgressIntentData {
  appType?: string
  keywords: string[]
  targetScenario?: string
}

export interface ProgressSourceCompletedData {
  platform?: string
  count?: number
}

export interface ProgressExtractionCompletedData {
  count?: number
}

export interface ProgressPlanningCompletedData {
  count?: number
  families: string[]
}

const PIPELINE_EVENT_TYPES: EventType[] = [
  'intent_started',
  'intent_parsed',
  'query_planning_started',
  'query_planning_completed',
  'source_started',
  'source_completed',
  'source_failed',
  'extraction_started',
  'extraction_completed',
  'aggregation_started',
  'aggregation_completed',
  'report_ready',
  'cancelled',
  'error',
]

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

function isEventType(value: unknown): value is EventType {
  return typeof value === 'string' && PIPELINE_EVENT_TYPES.includes(value as EventType)
}

function readString(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim().length > 0 ? value : undefined
}

function readStringList(value: unknown): string[] {
  if (!Array.isArray(value)) return []
  return value.filter((item): item is string => typeof item === 'string' && item.trim().length > 0)
}

function readCount(value: unknown): number | undefined {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value
  }
  if (typeof value === 'string' && value.trim().length > 0) {
    const parsed = Number(value)
    if (Number.isFinite(parsed)) {
      return parsed
    }
  }
  return undefined
}

export function parseIntentProgressData(raw: unknown): ProgressIntentData {
  if (!isRecord(raw)) {
    return { keywords: [] }
  }

  return {
    appType: readString(raw.app_type),
    keywords: readStringList(raw.keywords),
    targetScenario: readString(raw.target_scenario),
  }
}

export function parseSourceCompletedProgressData(raw: unknown): ProgressSourceCompletedData {
  if (!isRecord(raw)) {
    return {}
  }

  return {
    platform: readString(raw.platform),
    count: readCount(raw.count),
  }
}

export function parseExtractionCompletedProgressData(raw: unknown): ProgressExtractionCompletedData {
  if (!isRecord(raw)) {
    return {}
  }

  return {
    count: readCount(raw.count),
  }
}

export function parsePlanningCompletedProgressData(raw: unknown): ProgressPlanningCompletedData {
  if (!isRecord(raw)) {
    return { families: [] }
  }

  return {
    count: readCount(raw.count),
    families: readStringList(raw.families),
  }
}

export function normalizePipelineEventData(type: EventType, raw: unknown): PipelineEventData {
  if (type === 'intent_parsed') {
    const parsed = parseIntentProgressData(raw)
    return {
      ...(parsed.appType ? { app_type: parsed.appType } : {}),
      ...(parsed.keywords.length > 0 ? { keywords: parsed.keywords } : {}),
      ...(parsed.targetScenario ? { target_scenario: parsed.targetScenario } : {}),
    }
  }

  if (type === 'source_completed') {
    const parsed = parseSourceCompletedProgressData(raw)
    return {
      ...(parsed.platform ? { platform: parsed.platform } : {}),
      ...(parsed.count !== undefined ? { count: parsed.count } : {}),
    }
  }

  if (type === 'query_planning_completed') {
    const parsed = parsePlanningCompletedProgressData(raw)
    return {
      ...(parsed.count !== undefined ? { count: parsed.count } : {}),
      ...(parsed.families.length > 0 ? { families: parsed.families } : {}),
    }
  }

  if (type === 'extraction_completed') {
    const parsed = parseExtractionCompletedProgressData(raw)
    return {
      ...(parsed.count !== undefined ? { count: parsed.count } : {}),
    }
  }

  return {}
}

export function parsePipelineEvent(raw: unknown, fallbackType?: EventType): PipelineEvent | null {
  if (!isRecord(raw)) {
    return null
  }

  const type = isEventType(raw.type) ? raw.type : fallbackType
  if (!type) {
    return null
  }

  return {
    type,
    stage: readString(raw.stage) ?? '',
    message: readString(raw.message) ?? '',
    data: normalizePipelineEventData(type, raw.data),
    timestamp: readString(raw.timestamp) ?? '',
  }
}

export interface Competitor {
  name: string
  links: string[]
  one_liner: string
  features: string[]
  pricing: string | null
  strengths: string[]
  weaknesses: string[]
  relevance_score: number
  relevance_kind: 'direct' | 'adjacent'
  source_platforms: Platform[]
  source_urls: string[]
}

export interface SourceResult {
  platform: Platform
  status: SourceStatus
  raw_count: number
  competitors: Competitor[]
  error_msg: string | null
  duration_ms: number
}

export interface SearchQuery {
  platform: Platform
  queries: string[]
}

export interface PainSignal {
  theme: string
  summary: string
  intensity: number
  frequency: number
  evidence_urls: string[]
  source_platforms: Platform[]
}

export interface CommercialSignal {
  theme: string
  summary: string
  intent_strength: number
  monetization_hint: string
  evidence_urls: string[]
  source_platforms: Platform[]
}

export interface WhitespaceOpportunity {
  title: string
  description: string
  target_segment: string
  wedge: string
  potential_score: number
  confidence: number
  supporting_evidence: string[]
}

export interface OpportunityScoreBreakdown {
  pain_intensity: number
  solution_gap: number
  commercial_intent: number
  freshness: number
  competition_density: number
  score: number
}

export interface ConfidenceMetrics {
  sample_size: number
  source_coverage: number
  source_success_rate: number
  source_diversity: number
  evidence_density: number
  recency_score: number
  degradation_penalty: number
  contradiction_penalty: number
  reasons: string[]
  freshness_hint: string
  score: number
}

export type EvidenceCategory =
  | 'competitor'
  | 'pain'
  | 'commercial'
  | 'migration'
  | 'whitespace'
  | 'market'

export interface EvidenceItem {
  title: string
  url: string
  platform: Platform | null
  snippet: string
  category: EvidenceCategory
  freshness_hint: string
  matched_query: string
  query_family: string
}

export interface EvidenceSummary {
  top_evidence: string[]
  evidence_items: EvidenceItem[]
  category_counts: Record<string, number>
  source_platforms: Platform[]
  freshness_distribution: Record<string, number>
  degraded_sources: Platform[]
  uncertainty_notes: string[]
}

export interface CostBreakdown {
  llm_calls: number
  llm_retries: number
  endpoint_failovers: number
  source_calls: number
  pipeline_latency_ms: number
  tokens_prompt: number
  tokens_completion: number
}

export interface LlmFaultToleranceMeta {
  fallback_used: boolean
  endpoints_tried: string[]
  last_error_class: string
}

export interface ReportMeta {
  llm_fault_tolerance: LlmFaultToleranceMeta
  quality_warnings: string[]
}

export interface Intent {
  keywords_en: string[]
  keywords_zh: string[]
  exact_entities: string[]
  comparison_anchors: string[]
  search_goal: string
  app_type: string
  target_scenario: string
  output_language: string
  search_queries: SearchQuery[]
  cache_key: string
}

export type RecommendationType = 'go' | 'caution' | 'no_go'

export interface ResearchReport {
  id: string
  query: string
  created_at: string
  updated_at: string
  intent: Intent
  recommendation_type: RecommendationType
  go_no_go: string
  market_summary: string
  pain_signals: PainSignal[]
  commercial_signals: CommercialSignal[]
  whitespace_opportunities: WhitespaceOpportunity[]
  opportunity_score: OpportunityScoreBreakdown
  competitors: Competitor[]
  differentiation_angles: string[]
  evidence_summary: EvidenceSummary
  confidence: ConfidenceMetrics
  source_results: SourceResult[]
  cost_breakdown: CostBreakdown
  report_meta: ReportMeta
}

export interface ReportListItem {
  id: string
  query: string
  created_at: string
  competitor_count: number
}

export interface PaginatedReportList {
  items: ReportListItem[]
  total: number
  limit: number | null
  offset: number
}

export type RuntimeStatus =
  | 'processing'
  | 'failed'
  | 'cancelled'
  | 'complete'
  | 'not_found'

export interface ReportRuntimeStatus {
  status: RuntimeStatus
  report_id: string
  error_code?: string | null
  message?: string | null
  updated_at?: string | null
  query?: string | null
}
